### PR TITLE
Support v4 mapping to v6

### PR DESCRIPTION
**Pull request by jukkar**
This PR introduces v4-mapping-to-v6 and properly implements IPV6_V6ONLY socket option.
The echo sample application is converted to use this feature.

### CODE REVIEW

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -78,6 +78,15 @@ bool net_context_is_reuseport_set(struct net_context *context)
 #endif
 }
 
+bool net_context_is_v6only_set(struct net_context *context)
+{
+#if defined(CONFIG_NET_IPV4_MAPPING_TO_IPV6)
+	return context->options.ipv6_v6only;
+#else
+	return true;
```
Comment (pdgendt) [inline subsys/net/ip/net_context.c:86]: Missing `ARG_UNUSED(context);`

File: subsys/net/ip/Kconfig [hunk 2]
```diff
@@ -143,6 +143,18 @@ source "subsys/net/ip/Kconfig.ipv6"
 
 source "subsys/net/ip/Kconfig.ipv4"
 
+config NET_IPV4_MAPPING_TO_IPV6
+	bool "Support IPv4 mapped on IPv6 addresses"
+	depends on NET_NATIVE_IPV6
+	help
+	  Support v4-mapped-on-v6 address type. This allows IPv4 and IPv6
+	  to share a local port space. When application gets an IPv4
```
Comment (pdgendt) [inline subsys/net/ip/Kconfig:151]:
```suggestion
	  to share a local port space. When the application gets an IPv4
```

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -1496,6 +1513,32 @@ static int get_context_reuseport(struct net_context *context,
...
+
+	if (*len != sizeof(int)) {
+		return -EINVAL;
+	}
+
+	if (context->options.ipv6_v6only == true) {
+		*((int *)value) = (int) true;
+	} else {
+		*((int *)value) = (int) false;
+	}
```
Comment (pdgendt) [inline subsys/net/ip/net_context.c:1532]:
```suggestion
	*((int *)value) = (int)context->options.ipv6_v6only;
```

File: subsys/net/ip/net_context.c [hunk 4]
```diff
@@ -1496,6 +1513,32 @@ static int get_context_reuseport(struct net_context *context,
...
+
+	*len = sizeof(int);
+
+	return 0;
+#else
+	return -ENOTSUP;
```
Comment (pdgendt) [inline subsys/net/ip/net_context.c:1538]: Missing `ARG_UNUSED(context);`, `ARG_UNUSED(value);` and `ARG_UNUSED(len);`

**Review by pdgendt — COMMENTED** [review]
Thanks for this! Some minor remarks after a short review.

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c:1532]: I decided not to do this in order this function look similar as the other get socket option functions.

**Comment by jukkar** [comment]
@kartben The documentation build complains about this comment in the code
```
 *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
 *         look like ::ffff:a.b.c.d
```
The warning is this one
```
WARNING: /runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:1258: explicit link request to 'ffff' could not be resolved
```
Is there a way to mark this in the code so that doxygen does not think it is a link?

File: include/zephyr/net/net_ip.h [hunk 5]
```diff
@@ -1239,6 +1239,39 @@ static inline void net_ipv6_addr_create_ll_allrouters_mcast(struct in6_addr *add
...
+						  struct in6_addr *addr6)
+{
+	net_ipv6_addr_create(addr6, 0, 0, 0, 0, 0, 0xffff,
+			     ntohs(addr4->s4_addr16[0]),
+			     ntohs(addr4->s4_addr16[1]));
+}
+
+/**
+ *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
+ *         look like ::ffff:a.b.c.d
```
Comment (kartben) [inline include/zephyr/net/net_ip.h:1258]: The issue is with Sphinx, that thinks it's a directive. The following should do the trick (and it's not a bad idea anyway to format the IP address as code, I guess?)

```suggestion
 *         look like `::ffff:a.b.c.d`
```

Comment (jukkar) [reply include/zephyr/net/net_ip.h]: I figured it out, using `\` worked
```
diff --git a/include/zephyr/net/net_ip.h b/include/zephyr/net/net_ip.h
index 20d95d68b8..bc760e7323 100644
--- a/include/zephyr/net/net_ip.h
+++ b/include/zephyr/net/net_ip.h
@@ -1255,7 +1255,7 @@ static inline void net_ipv6_addr_create_v4_mapped(const struct in_addr *addr4,
 
 /**
  *  @brief Is the IPv6 address an IPv4 mapped one. The v4 mapped addresses
- *         look like `::ffff:a.b.c.d`
+ *         look like \::ffff:a.b.c.d
  *
  *  @param addr IPv6 address
  *
```

File: subsys/net/ip/net_context.c [hunk 6]
```diff
@@ -1726,8 +1775,21 @@ static int context_sendto(struct net_context *context,
 
 	} else if (IS_ENABLED(CONFIG_NET_IPV4) &&
 		   net_context_get_family(context) == AF_INET) {
```
Comment (rlubos) [inline subsys/net/ip/net_context.c:1777]: Is this the right place for the translation? This part of code is covering IPv4 sockets (AF_INET), I had an impression those should still cover IPv4 communications only? Shoudn't the IPv6 code (above) check for IPv4 mapped address?

File: tests/net/socket/misc/src/main.c [hunk 7]
```diff
@@ -610,4 +864,11 @@ ZTEST_USER(socket_misc_test_suite, test_ipv6)
 	test_ipv6_getpeername();
 }
 
+ZTEST_USER(socket_misc_test_suite, test_ipv4_mapped_to_ipv6)
+{
+	test_ipv4_mapped_to_ipv6_disabled();
+	test_ipv4_mapped_to_ipv6_enabled();
+	test_ipv4_mapped_to_ipv6_server();
```
Comment (rlubos) [inline tests/net/socket/misc/src/main.c:872]: Do those have to execute sequentially, i. e. couldn't they be separate test cases?

Comment (jukkar) [reply tests/net/socket/misc/src/main.c:872]: They could be separate as they do not depend on each other. I can change this if needed.

File: subsys/net/ip/net_context.c (hunk 6 above)
Comment (jukkar) [reply subsys/net/ip/net_context.c:1777]: When the socket is accepted and if the incoming data is IPv4, then the accepted net_context family is set to IPv4 even if the listening one is IPv6 and the application socket is IPv6. It looks like a big mess if we try to handle IPv4 data in IPv6 path in other parts of the network stack. With these minimal changes, the echo sample application I modified works fine both in Zephyr and Linux.

File: subsys/net/ip/connection.c [hunk 8]
```diff
@@ -380,6 +380,12 @@ int net_conn_register(uint16_t proto, uint8_t family,
 
 	conn_set_used(conn);
 
+	if (net_context_is_v6only_set(context)) {
+		conn->v6only = true;
```
Comment (rlubos) [inline subsys/net/ip/connection.c:384]: nit: Maybe just assign the result directly?
```
conn->v6only = net_context_is_v6only_set(context);
```

File: subsys/net/ip/connection.c [hunk 9]
```diff
@@ -740,7 +756,24 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
...
+					      !conn->v6only &&
+					      net_ipv6_is_addr_unspecified(
+						      &net_sin6(&conn->local_addr)->sin6_addr))) {
+						continue; /* wrong local address */
+					}
+				} else {
+					continue; /* wrong local address */
+				}
+
+				/* We might have a match for v4-to-v6 mapping, check even more */
```
Comment (rlubos) [inline subsys/net/ip/connection.c:776]: I'm not sure I understand this comment, `check even more` what?

**Review by rlubos — COMMENTED** [review]
Any clue what's wrong with the new testcase?

I must say this mapping seems to me as a bit tricky topic, but on the other side the echo sample seems to work fine now with both IPv4 and IPv6 and we get some extra test coverage, so fine by me.

File: subsys/net/ip/connection.c (hunk 9 above)
Comment (jukkar) [reply subsys/net/ip/connection.c:776]: Heh, that was a leftover comment from when testing the connection logic, I will tweak this to be more sane.

**Comment by jukkar** [comment]
The net-shell refactoring PR https://github.com/zephyrproject-rtos/zephyr/pull/63164 should be merged before this one so adding DNM

File: subsys/net/ip/net_context.c [hunk 10]
```diff
@@ -2591,6 +2653,28 @@ static int set_context_reuseport(struct net_context *context,
...
+
+	if (*((int *) value) != 0) {
+		v6only = true;
+	}
+
+	context->options.ipv6_v6only = v6only;
+
+	return 0;
+#else
+	return -ENOTSUP;
```
Comment (pdgendt) [inline subsys/net/ip/net_context.c:2674]:
```suggestion
	ARG_UNUSED(context);
	ARG_UNUSED(value);
	ARG_UNUSED(len);

	return -ENOTSUP;
```

**Comment by kartben** [comment]
@jukkar It feels like the README for the sample should be updated?

**Review by pdgendt — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**